projects/wallet-service-reliable-integration/0001-design.md

```diff
@@ -0,0 +1,205 @@
+**Transitions**:
+- If the service invocation completes successfully (`onDone`):
+  - **Target**: The machine transitions to the `success` sub-state.
+- If there's an error during the service invocation (`onError`):
+  - **Target**: The machine transitions to the `error` sub-state.
+
+
+#### `CONNECTED.handlingMetadataChanged`
+
+The machine will transition to this state to process a `NEW_METADATA_ACCEPTED` event.
+
+This event happens when the metadata of a transaction changes, this can happen for many reasons, but the ones we care about are:
+
+- The transaction has been confirmed by a block
+- The transaction was voided
+
+**Services**:    
+- The machine invokes (on entry) the `handleMetadataChanged` service, passing the event data. This will be further explained in the `Services` section of this document.
+**Transitions**:
+- If the service invocation completes successfully (`onDone`):
+    - **Target**: The machine transitions to the `CONNECTED.success` sub-state.
+- If there's an error during the service invocation (`onError`):
+    - **Target**: The machine transitions to the `CONNECTED.error` sub-state.
+
+
+#### `CONNECTED.success`
+
+The machine will transition to this state when a event is successfully processed
+
+**Entry Action**: As soon as the machine enters the `success` state, it sends an acknowledgment using the `sendAck` action.
+**Transitions**:
+  - The machine always transitions back to the `CONNECTED.idle` sub-state after performing the entry actions.
+
+
+#### `CONNECTED.error`
+
+The machine will transition to this state when an error is detected when event handling or invoking a `service`.
+
+This is a `final` state, meaning that no further transitions can occur from this state and that the machine will need manual intervention to recover.
+
+**Entry Action**: The machine logs the error using the `logError` action.
+
+
+## Potential challenges
+
+1. **Event Ordering and Consistency**: Given that events are sent in the order they occurred, it's crucial to ensure that the daemon maintains this order when processing events. This is especially important in scenarios like reorgs where the event order can significantly impact the wallet-service's state.
+2. **Event Volume and Scalability**: Depending on the full-node's activity, the number of events generated can be significant and considering that we are processing those events one at a time, we should make sure that we are doing this as efficiently as possible
+3. **Error Handling and Recovery**: There might be situations where the daemon encounters an error processing an event or loses connection to the full-node. We should make sure that we always have the latest successfully synced event id persisted and that we are rolling back changes that were unsuccessful to prevent invalid states
+
+
+# Guide-level explanation
```

Comment:
After a connection is established, you are supposed to send the `id` of the next event you expect to receive. I couldn't find where you are persisting the next event `id` and the `peer-id` in case the sync daemon dies.

Reply:
Sorry, added the service that will store the `event_id` in the `success` state

b2662064da15bf5ccc261f0fcef9c82aaa3ba49b

Reply:
Great! Thanks!